index.js

```diff
@@ -618,27 +616,24 @@ module.exports = class Hypercore extends EventEmitter {
     const activeRequests = (opts && opts.activeRequests) || this.activeRequests
     const req = this.replicator.addUpgrade(activeRequests)
 
-    if (this._preupgrade && this._preupgradeLength === null) {
-      this._preupgradeLength = this.core.tree.length
-    }
+    if (this._preupgrade) this._updateSnapshot()
 
     let upgraded = await req.promise
 
     if (this._preupgrade) {
-      if (this._postupgradeLength === null) {
+      if (this._preupgrading === null) {
         const latest = this.session()
 
-        this._postupgradeLength = Promise.resolve(this._preupgrade(latest))
-        this._postupgradeLength
+        this._preupgrading = Promise.resolve(this._preupgrade(latest))
+        this._preupgrading
           .catch(noop)
           .then(() => latest.close())
       }
 
-      const length = await this._postupgradeLength
-      const preupgradeLength = this._preupgradeLength
+      const length = await this._preupgrading
+      const preupgradeLength = this._snapshot.length
 
-      this._preupgradeLength = null
-      this._postupgradeLength = null
+      this._preupgrading = null
```

Comment:
i think this is only safe if you save the promise first to a var, pre-wait and then check that's still the same var.

ie

```
const preupd = this._preupgrading
await this._preupgrading
...
if (this._preupgrading === preupd) this._preupgrading = null
```

Otherwise someone might have scheduled another preupgrade inbetween the await